### ISSUE TITLE
diy::io::write_blocks  lock when nbblock % nbproc != 0

### ISSUE THREAD

**Issue by mrzv** [diatomic/diy2#30]
**Issue by [Hadrien Croubois](https://bitbucket.com/hadriencroubois)**
_Monday Mar 02, 2015 at 22:21 GMT_

----

I noticed locks in diy::io::write_blocks when the number of block is not constant among the different processes.

**Comment by mrzv** [comment]
**Comment by [Hadrien Croubois](https://bitbucket.com/hadriencroubois)**
_Monday Mar 02, 2015 at 22:48 GMT_

----

include/diy/io/block.hpp --line 88

```
#!C++

{
++      mpi::all_reduce(comm, count, shift, std::plus<offset_t>());
        // -1 indicates that there is no block written here from this rank
        offset_counts.push_back(GidOffsetCount(-1, offset, count));
}
```

basically all processes were not taking part of the all_reduce. This simple line works as it will add a 0 contribution (count = 0, line 66) to shift, hense avoiding the deadlock without modifying the result of the mpi::all_reduce collective

**Comment by mrzv** [comment]
**Comment by [Dmitriy Morozov](https://bitbucket.com/mrzv)**
_Monday Mar 02, 2015 at 23:23 GMT_

----

It also needs a matching `mpi::scan()`, I think.

**Comment by mrzv** [comment]
**Comment by [Dmitriy Morozov](https://bitbucket.com/mrzv)**
_Monday Mar 02, 2015 at 23:24 GMT_

----

Added missing matching MPI global operations in diy::io::write_blocks() (fixes #30)


→ f73ab3472cbf
